### ISSUE TITLE
maven-publish: using loom, components.java exposes way too many dependencies

### ISSUE THREAD

**Issue by falkreon** [FabricMC/fabric-loom#49]
Currently, just dropping maven-publish into a fabric gradle project creates a crazed pom with literally every dependency of minecraft, plus fabric, plus every modCompile. The input to maven-publish is usually `components.java`, and that is generally the only component that ever exists.

Right now there's a workaround involving going into the xml and literally deleting the dependencies node, but this.... isn't ideal.

**Comment by NikkyAI** [comment]
what would be the preferred solution?
fabric could provide its own named `SoftwareComponent` so that you just need to configure `from(components.modDependencies)` in the publication
this could be possible

**Comment by falkreon** [comment]
I like providing a SoftwareComponent. It's just a matter of getting the right dependencies into it, for instance the author might want to declare some of the items normally declared modCompile, but not others. Maybe a configuration that goes straight into the component?

**Comment by NikkyAI** [comment]
you would want dependencies on other mods to go in but not loader generally..
unless you jsut go ahead and include which version of `fabric-loader` you used as a dependency and whoever consumes the library will use that version unless they specify their own version of `fabric-loader`

i would also support a configuration `modApi` or `apiMod` that is `modCompile` and goes into the dependencies component

**Comment by UpcraftLP** [comment]
>i would also support a configuration `modApi` or `apiMod` that is `modCompile` and goes into the dependencies component

maybe `modApi` and `modImplementation`, and deprecate `modCompile`? gradle did this a while ago to the `compile` configuration: It has been split into `api` and `implementation` and is now deprecated.

**Comment by NikkyAI** [comment]
but does it make sense for mod dependencies? what is the difference between api and implementation and do we need that difference with mod dependencies ?

**Comment by Mordna** [comment]
'api' is basically like the old 'compile' meaning that any 'api' components (recursively) can be used/called.
'implementation' allows only usage of the stated library/mod without the ability to directly call/use its dependencies.
[This](https://stackoverflow.com/a/44419574) explains it way better.

**Comment by NikkyAI** [comment]
i think everyone understand what `implementation` does, ot at least i hope so

the question is.. does it make sense here? where would zou want to use a `modImplementation` and not add that to the dependencies block in the pom?
maybe when using jar-in-jar you might want to only expose that you are using `API A`

apart from that i see no usecase for `modImplementation`

**Comment by JamiesWhiteShirt** [comment]
There is a separate issue about more choices of configurations, #57 . As far as I can tell the scope of this issue is the amount of dependencies unexpectedly appearing in the POM.

**Comment by falkreon** [comment]
Whatever changed in loom from 0.2.0->0.2.1 is now preventing publish, period.

Not removing any dependencies results in two `<dependencies/>` blocks: One for all the junk including asm, patchy, javabridge, etc.; the other for fabric-loom and fabric api.

Some manual poking at task configuration time suggests that this second dependencies block is added after configuration time, which is a Very Bad Thing - having wrong behavior before configuration time is better than having correct behavior after configuration time. At least then when circumstances change the user can step in and create workarounds.

So I guess it's PR time. I'll repro the problem and then get cracking on a fix.

**Comment by falkreon** [comment]
Update: Couldn't repro on exampleMod (though I confirmed that the dependency injection happens far too late), so the main problem remaining is that the remapJar task only declares the '-dev' jar as its output, preventing the actual remapped output from properly being captured and used in some publish contexts (try using shadow, for instance; you'll wind up publishing a non-remapped jar!)

The workaround is to manually specify the files, like:
```groovy
artifact ("${project.buildDir.absolutePath}/libs/${archivesBaseName}-${project.version}-shadow.jar") {
	classifier null
	builtBy remapJar
}
```

Guess this is straying from the main topic now, and as long as one hews close to examplemod's pattern, this issue appears to be..... fixed?!? in 0.2.2
